arangod/Graph/Providers/ClusterProvider.cpp

```diff
@@ -130,28 +130,49 @@ auto ClusterProvider<StepImpl>::startVertex(const VertexType& vertex,
 template<class StepImpl>
 void ClusterProvider<StepImpl>::fetchVerticesFromEngines(
     std::vector<Step*> const& looseEnds, std::vector<Step*>& result) {
-  auto const* engines = _opts.engines();
   // slow path, sharding not deducable from _id
+  bool mustSend = false;
+
   transaction::BuilderLeaser leased(trx());
   leased->openObject();
-  leased->add("keys", VPackValue(VPackValueType::Array));
-  for (auto const& looseEnd : looseEnds) {
-    auto const& vertexId = looseEnd->getVertex().getID();
-    if (!_opts.getCache()->isVertexCached(vertexId)) {
-      leased->add(VPackValuePair(vertexId.data(), vertexId.length(),
-                                 VPackValueType::String));
+
+  if (_opts.produceVertices()) {
+    // slow path, sharding not deducable from _id
+    leased->add("keys", VPackValue(VPackValueType::Array));
+    for (auto const& looseEnd : looseEnds) {
+      TRI_ASSERT(looseEnd->isLooseEnd());
+      auto const& vertexId = looseEnd->getVertex().getID();
+      if (!_opts.getCache()->isVertexCached(vertexId)) {
+        leased->add(VPackValuePair(vertexId.data(), vertexId.size(),
+                                   VPackValueType::String));
+        mustSend = true;
+      }
     }
+    leased->close();  // 'keys' Array
   }
-  leased->close();  // 'keys' Array
   leased->close();  // base object
 
+  if (!mustSend) {
+    // nothing to send. save requests...
+    for (auto& looseEnd : looseEnds) {
+      auto const& vertexId = looseEnd->getVertex().getID();
+      if (!_opts.getCache()->isVertexCached(vertexId)) {
+        _opts.getCache()->cacheVertex(vertexId, VPackSlice::nullSlice());
+      }
+      result.emplace_back(looseEnd);
+      looseEnd->setVertexFetched();
+    }
+    return;
+  }
+
   auto* pool =
       trx()->vocbase().server().template getFeature<NetworkFeature>().pool();
 
   network::RequestOptions reqOpts;
   reqOpts.database = trx()->vocbase().name();
   reqOpts.skipScheduler = true;  // hack to avoid scheduler queue
 
+  auto const* engines = _opts.engines();
   std::vector<Future<network::Response>> futures;
   futures.reserve(engines->size());
```

Comment:
Same suggestion already mentioned here: https://github.com/arangodb/arangodb/pull/17547#discussion_r1019008799

Reply:
The `engines` variables is also used further down, so we cannot remove it.